File: pi_image/sensors/interfacing/windVane/ADC.py
'''
  ADC.py

  By: Joey-Michael Fallone and Dallin Toth

  Uses the existing Adafruit_ADS1x15 library to print out a 
  ADC value read to stdout

  This code is based on examples from the library referenced, 
  for which license information is provided in the legal section.
'''

# constants
GAIN = 1 # 1 is used on adc datasheet
# this value should be changed for 3.3 V operation in the future
# this value is for 4.09v operation

import Adafruit_ADS1x15
adc = Adafruit_ADS1x15.ADS1115()
values = adc.read_adc(0, gain=GAIN)
#print('{0:>6}'.format(*values))
print(values)

File: pi_image/sensors/interfacing/WindVane/ADC.cpp
/*
 * ADC.cpp
 *
 * By: Joey-Michael Fallone and Dallin Toth
 * ADC Wrapper fro python ADC Lib
 *
 * This simple wrapper uses os calls to read ADC values in over I2C 
 * using the existing python lib for this ADC on raspberry pi
 *
 *
*/

#include "ADC.h"
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <pthread.h>
// constructor
ADC::ADC () {
  ; // for now no init steps
}
/*
 * ADC::readVoltage()
 *
 * Reads in the voltage value and returns it as a float
 * between 0 and 3.3 V
 *
 *
 *
*/
float ADC::readVoltage() {
  // based on stackoverflow.com/questions/17071702/c-language-read-from-stdout


	int stdout_bk; // backup stdout

	stdout_bk = dup(fileno(stdout));
	int pipefd[2];
	pipe2(pipefd, NON_BLOCKING);

	dup2(pipefd[1], fileno(stdout));
	system("python ADC.py");
	fflush(stdout);
	close(pipefd[1]);

	dup2(stdout_bk, fileno(stdout));

	char buff[50];

	read(pipefd[0], buff, 49);
	return ADC::convertReadingToVoltage(atoi(buff));

}

float ADC::convertReadingToVoltage(int reading) {
  return ((reading*VDD)/MAX_READ);
}
